File: C#-Advanced/C#-OOP/04.PolymorphismExercise/WildFarm/Animals/Animal.cs
using System;
using System.Collections.Generic;
using System.Text;
using WildFarm.Foods;

namespace WildFarm.Animals
{
    public abstract class Animal
    {
        public Animal(string name, double weight, HashSet<string> allowedFoods, double weightModifier)
        {
            Name = name;
            Weight = weight;
            AllowedFoods = allowedFoods;
            WeightModifier = weightModifier;
        }

        private HashSet<string> AllowedFoods { get; set; }

        public double WeightModifier { get; private set; }

        public string Name { get; private set; }

        public double Weight { get; protected set; }

        public int FoodEaten { get; protected set; }

        public virtual string ProduceSound()
        {
            return string.Empty;
        }

        public virtual void Eat(Food food)
        {
            if (AllowedFoods.Contains(food.GetType().Name))
            {
                FoodEaten += food.Quantity;
                Weight += food.Quantity * WeightModifier;
            }
            else
            {
                throw new InvalidOperationException($"{GetType().Name} does not eat {food.GetType().Name}!");
            }
        }
    }
}

File: Python-OOP/06Polymorphism-and-Abstraction-Exercise/04Wild-Farm/project/animals/animal.py
from abc import ABC, abstractmethod


class Animal(ABC):
    SOUND = ''
    ALLOWED_FOODS = []
    WEIGHT_GAIN = 0

    @abstractmethod
    def __init__(self, name, weight):
        self.name = name
        self.weight = weight
        self.food_eaten = 0

    def make_sound(self):
        return self.SOUND

    def feed(self, food):
        food_name = food.__class__.__name__

        if food_name not in self.ALLOWED_FOODS:
            return f"{self.__class__.__name__} does not eat {food_name}!"

        self.weight += self.WEIGHT_GAIN * food.quantity
        self.food_eaten += food.quantity


class Bird(Animal, ABC):
    @abstractmethod
    def __init__(self, name, weight, wing_size):
        super().__init__(name, weight)
        self.wing_size = wing_size

    def __str__(self):
        return f'{self.__class__.__name__} [{self.name}, {self.wing_size}, {self.weight}, {self.food_eaten}]'


class Mammal(Animal, ABC):
    @abstractmethod
    def __init__(self, name, weight, living_region):
        super().__init__(name, weight)
        self.living_region = living_region

    def __str__(self):
        return f"{self.__class__.__name__} [{self.name}, {self.weight}, {self.living_region}, {self.food_eaten}]"
